Don't retransmit lost packets that are acked later (#3957)
quicwg/base-drafts: draft-ietf-quic-transport.md

packets containing an outdated frame, such as a MAX_DATA frame carrying a
smaller maximum data than one found in an older packet.

A sender SHOULD avoid retransmitting information from packets once they are
acknowledged. This includes packets that are acknowledged after being declared
lost. Doing so requires senders to retain information about packets after they

> are declared lost. A sender can discard this information after a period of time
> elapses, such as three times the PTO (Section 6.2 of {{QUIC-RECOVERY}}), or on

Reviewer: ```suggestion
are declared lost. A sender can discard this information a period of time after the
packet is lost, such as after the Smoothed RTT (Section 5 of {{QUIC-RECOVERY}}), or on
```

Author: I'm happy to change it to 1 x PTO. I'd like the ack delay and the rttdev to be included in this suggestion.

Reviewer: ack delay shouldn't need to be included, because the packet would be acknowledged immediately.

Including RTTVar implies that the network's reordering is sometimes as much 3*SRTT, but assuming the data was retransmitted, you'd expect the retransmission to be acknowledged in about 1 RTT, so suggesting you might need to wait longer seems odd to me.

> acknowledged. This includes packets that are acknowledged after being declared
> lost. Doing so requires senders to retain information about packets after they
> are declared lost. A sender can discard this information after a period of time
> elapses, such as a PTO (Section 6.2 of {{QUIC-RECOVERY}}), or on other events,
> such as reaching a memory limit.

Reviewer: ```suggestion
acknowledged. Doing this for packets acknowledged after they are declared
lost requires retaining state for lost packets. A sender can discard this information
after the expected maximum packet reordering elapses, such as after PTO has elapsed
(Section 6.2 of {{QUIC-RECOVERY}}), or on other events, such as reaching a packet limit.
```

Author: I will take the compression, but we don't need to say "expected maximum packet reordering", since that's not a well-defined term. I'll spell in out in other words.

Author: I'm trying to wordsmith this compression, but it's a bit difficult. The text is a bit verbose, but I think it is clearer. Unless you think there's a substantive difference, I'd prefer the existing text for clarity.

Reviewer 2: The expanded form is fine.

Reviewer: The intent was to both make it clearer and talk about the goal of waiting for a reordering period.  I find the existing text a bit unclear, which is why I put the first two sentences together.

Author: Ok, added some rationale.

packets containing an outdated frame, such as a MAX_DATA frame carrying a
smaller maximum data than one found in an older packet.

A sender SHOULD avoid retransmitting information from packets once they are

> acknowledged. This includes packets that are acknowledged after being declared
> lost, which can happen in the presence of network reordering. Doing so requires
> senders to retain information about packets after they are declared lost. A

Reviewer: ```suggestion
acknowledged. This includes packets acknowledged after being
declared lost due to network reordering, which requires senders to
retain information about packets after they are declared lost. A
```

Author: That's a run-on sentence.

Reviewer: Fair, I just find the current structure awkward.  Either way, this proposal is basically done.

they are declared lost. A sender can discard this information after a period of

> time elapses, such as three times the PTO (Section 6.2 of {{QUIC-RECOVERY}}), or

Reviewer: 3*PTO is really long in this case, and it can consume a fair bit of memory depending upon implementation details.  I'd suggest an ~RTT, since by that point you're expecting the acknowledgement of the retransmitted data.

Reviewer 3: Generally speaking, I tend to think that 3*PTO is fine, and that is because the size of a sentmap entry (i.e., the amount of memory required to remember what was sent in a packet) is only a tiny fraction of the MTU.

If your send buffer is capable of retaining X packets amount of data, then it does not matter if you use *one* tiny fraction, *two* tiny fractions,  or *three*, for sentmap.

All that said, I agree that it might sound a bit frightening, it does not need to be as long as 3*PTO, and therefore my view is that this is a trade-off issue between simplicity (i.e. reusing 3*PTO timer) vs. less frightening.

> When a sender declares a packet as lost and either retransmits the contained
> frames or marks them for retransmission, the sender SHOULD avoid subsequent
> retransmission of this information if it later receives an acknowledgement for
> that packet. Doing so requires senders to retain information about packets after
> they are declared lost. A sender can discard this information after a period of
> time elapses, such as three times the PTO (Section 6.2 of {{QUIC-RECOVERY}}), or
> on other events, such as reaching a memory limit.

Reviewer 2: ```suggestion
Acknowledgments can be received after a sender declares a packet as lost.  A
sender SHOULD avoid sending information again after receiving a late
acknowledgment. Doing so requires senders to retain information about packets
for some time after they are declared lost. A sender can discard this
information after a period of time elapses, such as three times the PTO (Section
6.2 of {{QUIC-RECOVERY}}), or on other events, such as reaching a memory limit.
```

Reviewer 3: I think @martinthomson's idea of splitting the sentence into two. The only thing I might prefer is changing the last part of the second sentence "after receiving a late acknowledgement" to something like "after receiving such an acknowledgement" to make it clearer, but anyways it's good.

Reviewer: How about "..after receiving an acknowledgement for a previously lost packet."  I think late acknowledgement is potentially confusing.

Reviewer: And how about starting with "A packet can be acknowledged after being declared lost." instead of "Acknowledgements can be received after a sender declares a packet as lost."
